File: studio/AuraStudio/Sources/AuraStudio/Services/FirmwareSwitcher.swift
import Foundation

/// ST-056 / CONTRATO-firmware-studio.md v10 (ST-065: hasta tres): varios firmwares instalados a la
/// vez y conmutacion entre ellos POR RENOMBRE, sin borrar ni descargar.
///
/// El arbol activo es siempre `/.rockbox/` (lo unico que el bootloader
/// compartido sabe arrancar); el de la otra familia duerme como
/// `/.firmware-aura/` o `/.firmware-metro/`, completo y con sus propios
/// ajustes. Cambiar de firmware son dos renombres en FAT (instantaneos),
/// en un orden que importa -- el saliente primero -- mas el respaldo del
/// bootloader en la raiz y el marcador de sync para que el entrante
/// reconstruya su base de datos. Todo son operaciones de FileManager
/// sobre el volumen montado: se prueba con carpetas temporales.
///
/// Esta misma secuencia la ejecuta el firmware desde Ajustes (M-090 /
/// D-327); aqui esta la version de Studio.
///
/// INVARIANTE (ST-069 / contrato v15): todo lo de aqui opera SOLO sobre
/// `/.rockbox/`, `/.firmware-*/` y `/rockbox.ipod`. `/.aura/tagcache/`
/// (base de datos compartida por los tres firmwares) y `/.aura/thumbs/`
/// (miniaturas compartidas) son propiedad del firmware aunque vivan bajo
/// `/.aura/`: ningun renombre, estacionamiento, reparacion, siembra ni
/// espejo los toca. Ver `LibrarySync.sharedTagcacheDirRelativePath`.
enum FirmwareSwitcher {
    static let activeTreeName = ".rockbox"
    static let rootFirmwareBinaryName = "rockbox.ipod"
    /// ST-059 / contrato v12: sello de biblioteca compartido (cambia solo
    /// cuando un sync toca musica) y su anotacion POR ARBOL (contra que
    /// sello construyo su base ese firmware). Igualdad exacta de sellos =
    /// la base del arbol entrante sigue valida = no se escribe marcador.
    static let libraryStampRelativePath = ".aura/library-stamp"
    /// Hasta v14 la anotacion vivia por arbol; desde v15 (ST-069) el
    /// firmware la lee/escribe en `/.aura/tagcache/db_stamp.txt`. Studio
    /// conserva la ruta por arbol solo para compararla al cambiar de
    /// familia con arboles anteriores a v15 y para borrarla al forzar una
    /// reconstruccion.
    static let dbStampRelativePathInTree = "aura/db_stamp.txt"

    enum SwitchError: Error, Equatable {
        /// No hay `/.firmware-<familia>/` que despertar.
        case dormantTreeMissing(FirmwareFamily)
        /// La familia pedida ya es la activa.
        case alreadyActive(FirmwareFamily)
        /// La familia activa no se pudo determinar y por lo tanto no hay
        /// nombre bajo el cual estacionarla.
        case activeFamilyUnknown
        /// Una familia que no se puede estacionar ni despertar.
        case familyNotSwitchable(FirmwareFamily)
    }

    /// Que familias tienen un arbol dormido en el volumen.
    static func dormantFamilies(volumeRoot: URL, fileManager: FileManager = .default) -> [FirmwareFamily] {
        FirmwareFamily.installable.filter { family in
            guard let name = family.dormantTreeName else { return false }
            var isDir: ObjCBool = false
            return fileManager.fileExists(atPath: volumeRoot.appendingPathComponent(name).path,
                                          isDirectory: &isDir) && isDir.boolValue
        }
    }

    static func hasActiveTree(volumeRoot: URL, fileManager: FileManager = .default) -> Bool {
        var isDir: ObjCBool = false
        return fileManager.fileExists(atPath: volumeRoot.appendingPathComponent(activeTreeName).path,
                                      isDirectory: &isDir) && isDir.boolValue
    }

    /// Un cambio que quedo a medias (bateria, cable): sin `/.rockbox/` pero
    /// con exactamente un arbol dormido. Se repara despertandolo. Con dos
    /// dormidos y ningun activo no se adivina: se deja como esta y se
    /// reporta `nil`.
    @discardableResult
    static func repairIfNeeded(volumeRoot: URL, fileManager: FileManager = .default) throws -> FirmwareFamily? {
        guard !hasActiveTree(volumeRoot: volumeRoot, fileManager: fileManager) else { return nil }
        let dormant = dormantFamilies(volumeRoot: volumeRoot, fileManager: fileManager)
        guard dormant.count == 1, let family = dormant.first, let name = family.dormantTreeName else { return nil }
        try fileManager.moveItem(at: volumeRoot.appendingPathComponent(name),
                                 to: volumeRoot.appendingPathComponent(activeTreeName))
        try refreshRootBinary(volumeRoot: volumeRoot, fileManager: fileManager)
        return family
    }

    /// Estaciona el arbol activo como dormido de `family` (reemplazando un
    /// dormido anterior de esa misma familia, si lo hubiera -- nunca dos
    /// de la misma). Lo usa el instalador antes de instalar la OTRA
    /// familia en `/.rockbox/`, en vez de borrarla.
    static func parkActiveTree(as family: FirmwareFamily, volumeRoot: URL,
                               fileManager: FileManager = .default) throws {
        guard let name = family.dormantTreeName else { throw SwitchError.familyNotSwitchable(family) }
        let active = volumeRoot.appendingPathComponent(activeTreeName)
        let dormant = volumeRoot.appendingPathComponent(name)
        guard fileManager.fileExists(atPath: active.path) else { return }
        if fileManager.fileExists(atPath: dormant.path) {
            try fileManager.removeItem(at: dormant)
        }
        try fileManager.moveItem(at: active, to: dormant)
    }

    /// Borra el arbol dormido de `family` si existe (tras instalar esa
    /// familia fresca en `/.rockbox/`: nunca un dormido de la familia
    /// activa).
    static func removeDormantTree(of family: FirmwareFamily, volumeRoot: URL,
                                  fileManager: FileManager = .default) throws {
        guard let name = family.dormantTreeName else { return }
        let dormant = volumeRoot.appendingPathComponent(name)
        if fileManager.fileExists(atPath: dormant.path) {
            try fileManager.removeItem(at: dormant)
        }
    }

    /// EL cambio (contrato v10, pasos 2-5; el 1 -- que el firmware guarde
    /// lo suyo -- no aplica desde Studio porque el firmware no esta
    /// corriendo en modo disco, y el 6 -- reiniciar -- lo hace el usuario
    /// al expulsar):
    ///   `/.rockbox/` -> `/.firmware-<activa>/`
    ///   `/.firmware-<destino>/` -> `/.rockbox/`
    ///   `/rockbox.ipod` := `/.rockbox/rockbox.ipod` del entrante
    ///   `/.aura/sync-pending.json` con music: true
    static func switchActiveFirmware(to target: FirmwareFamily, currentlyActive: FirmwareFamily,
                                     volumeRoot: URL, fileManager: FileManager = .default) throws {
        guard target != currentlyActive else { throw SwitchError.alreadyActive(target) }
        guard let targetName = target.dormantTreeName else { throw SwitchError.familyNotSwitchable(target) }
        guard let activeName = currentlyActive.dormantTreeName else { throw SwitchError.activeFamilyUnknown }

        let active = volumeRoot.appendingPathComponent(activeTreeName)
        let targetDormant = volumeRoot.appendingPathComponent(targetName)
        let parkedActive = volumeRoot.appendingPathComponent(activeName)

        guard fileManager.fileExists(atPath: targetDormant.path) else {
            throw SwitchError.dormantTreeMissing(target)
        }

        // (2) saliente primero: el peor caso (corte aqui) deja un arbol
        // dormido entero y ningun activo -- repairIfNeeded() lo levanta.
        if fileManager.fileExists(atPath: active.path) {
            if fileManager.fileExists(atPath: parkedActive.path) {
                try fileManager.removeItem(at: parkedActive)
            }
            try fileManager.moveItem(at: active, to: parkedActive)
        }
        // (3) entrante
        try fileManager.moveItem(at: targetDormant, to: active)
        // (4) respaldo del bootloader
        try refreshRootBinary(volumeRoot: volumeRoot, fileManager: fileManager)
        // (5) v12/ST-059: el marcador solo si la biblioteca cambio desde
        // que el ENTRANTE construyo su base -- sin sync de por medio, el
        // cambio es instantaneo y sin reconstruccion (reporte del dueño:
        // cada ida y vuelta costaba ~5 min de "optimizando").
        if !incomingDatabaseIsCurrent(volumeRoot: volumeRoot, outgoing: currentlyActive,
                                      fileManager: fileManager) {
            try SyncPendingMarker(changes: .init(music: true, video: false, images: false))
                .write(to: volumeRoot, fileManager: fileManager)
        }
    }

    /// ST-059: compara el sello del arbol ya renombrado a `/.rockbox/`
    /// (el entrante) contra `/.aura/library-stamp`. Si el sello
    /// compartido falta (primer cambio tras v12), lo crea y lo anota como
    /// del SALIENTE -- su base si esta al dia, acaba de estar corriendo --
    /// para que el proximo cambio de vuelta ya no reconstruya.
    private static func incomingDatabaseIsCurrent(volumeRoot: URL, outgoing: FirmwareFamily,
                                                  fileManager: FileManager) -> Bool {
        let stampURL = volumeRoot.appendingPathComponent(libraryStampRelativePath)
        var stamp = (try? String(contentsOf: stampURL, encoding: .utf8))?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if stamp == nil || stamp?.isEmpty == true {
            let fresh = Self.makeLibraryStamp()
            try? fileManager.createDirectory(at: stampURL.deletingLastPathComponent(),
                                             withIntermediateDirectories: true)
            try? (fresh + "\n").write(to: stampURL, atomically: true, encoding: .utf8)
            if let name = outgoing.dormantTreeName {
                let outgoingStamp = volumeRoot.appendingPathComponent(name)
                    .appendingPathComponent(dbStampRelativePathInTree)
                try? (fresh + "\n").write(to: outgoingStamp, atomically: true, encoding: .utf8)
            }
            stamp = fresh
        }
        let incoming = volumeRoot.appendingPathComponent(activeTreeName)
            .appendingPathComponent(dbStampRelativePathInTree)
        guard let recorded = try? String(contentsOf: incoming, encoding: .utf8) else { return false }
        return recorded.trimmingCharacters(in: .whitespacesAndNewlines) == stamp
    }

    static func makeLibraryStamp(date: Date = Date()) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f.string(from: date) + "-" + UUID().uuidString.prefix(8)
    }

    /// ST-059: Studio renueva el sello en cada sync que toca musica -- es
    /// LA definicion de "la biblioteca cambio" para el cambio de firmware.
    static func bumpLibraryStamp(volumeRoot: URL, fileManager: FileManager = .default) {
        let url = volumeRoot.appendingPathComponent(libraryStampRelativePath)
        try? fileManager.createDirectory(at: url.deletingLastPathComponent(),
                                         withIntermediateDirectories: true)
        try? (makeLibraryStamp() + "\n").write(to: url, atomically: true, encoding: .utf8)
    }

    /// `/rockbox.ipod` (raiz) = el binario del arbol activo. Es lo que el
    /// bootloader arranca si `/.rockbox/rockbox.ipod` falta; tiene que ser
    /// SIEMPRE el del firmware activo, nunca el de un arbol dormido.
    static func refreshRootBinary(volumeRoot: URL, fileManager: FileManager = .default) throws {
        let source = volumeRoot.appendingPathComponent(activeTreeName).appendingPathComponent(rootFirmwareBinaryName)
        let root = volumeRoot.appendingPathComponent(rootFirmwareBinaryName)
        guard fileManager.fileExists(atPath: source.path) else { return }
        if fileManager.fileExists(atPath: root.path) {
            try fileManager.removeItem(at: root)
        }
        try fileManager.copyItem(at: source, to: root)
    }

    /// Archivos del contrato que Studio escribe en `.rockbox/aura/` y que
    /// el firmware dormido tambien necesita al despertar (v10). Se copian
    /// tal cual desde el activo a cada dormido presente, al final de cada
    /// sync. `aura.cfg` NO (es de cada firmware); `themes/` tampoco (es de
    /// Aura, activo o dormido, y viaja con su arbol).
    /// NOTA (v12/ST-059): `aura/db_stamp.txt` tampoco se espeja -- es la
    /// anotacion de CADA arbol sobre su propia base de datos.
    /// NOTA (v11/ST-058): `aura/install_manifest.cfg` NO esta aqui a
    /// proposito -- es POR ARBOL (describe lo instalado en ESE arbol) y
    /// espejarlo haria que la actualizacion selectiva del dormido diera
    /// por escritos archivos que nunca se le escribieron.
    static let mirroredContractEntries = [
        "aura/sync_summary.cfg",
        "aura/sync_manifest.json",
        "aura/artist_images.cfg",
        "aura/artists",
        "aura/video_categories.cfg",
        "aura/photo_categories.cfg",
        "aura/ratings.cfg",
        "aura/device.cfg",
    ]

    /// ST-061: el problema inverso al espejo -- un arbol ACTIVO recien
    /// instalado (extraccion fresca) no tiene los archivos del contrato
    /// que Studio escribio en su vida anterior: sin `sync_summary.cfg`
    /// el firmware dice "sin sincronizar todavia", sin
    /// `artist_images.cfg`/`artists/` no hay fotos de artista, sin
    /// `*_categories.cfg` los videos/fotos salen sin clasificar (reporte
    /// del dueño con Metro). Si el arbol dormido SI los tiene (los
    /// conservo al estacionarse, o el espejo de cada sync lo mantuvo al
    /// dia), se copian de ahi. Solo cuando el activo NO tiene
    /// `sync_summary.cfg` -- si lo tiene, lo suyo es igual o mas nuevo.
    @discardableResult
    static func seedContractFilesToActiveTree(volumeRoot: URL, fileManager: FileManager = .default) -> Bool {
        let active = volumeRoot.appendingPathComponent(activeTreeName)
        guard !fileManager.fileExists(atPath: active.appendingPathComponent("aura/sync_summary.cfg").path) else {
            return false
        }
        guard let donorFamily = dormantFamilies(volumeRoot: volumeRoot, fileManager: fileManager).first(where: { family in
            guard let name = family.dormantTreeName else { return false }
            return fileManager.fileExists(atPath: volumeRoot.appendingPathComponent(name)
                .appendingPathComponent("aura/sync_summary.cfg").path)
        }), let donorName = donorFamily.dormantTreeName else { return false }

        let donor = volumeRoot.appendingPathComponent(donorName)
        try? fileManager.createDirectory(at: active.appendingPathComponent("aura"),
                                         withIntermediateDirectories: true)
        for entry in mirroredContractEntries {
            let src = donor.appendingPathComponent(entry)
            let dst = active.appendingPathComponent(entry)
            guard fileManager.fileExists(atPath: src.path) else { continue }
            if fileManager.fileExists(atPath: dst.path) {
                try? fileManager.removeItem(at: dst)
            }
            try? fileManager.copyItem(at: src, to: dst)
        }
        return true
    }

    static func mirrorContractFilesToDormantTrees(volumeRoot: URL, fileManager: FileManager = .default) throws {
        let active = volumeRoot.appendingPathComponent(activeTreeName)
        for family in dormantFamilies(volumeRoot: volumeRoot, fileManager: fileManager) {
            guard let name = family.dormantTreeName else { continue }
            let dormant = volumeRoot.appendingPathComponent(name)
            try fileManager.createDirectory(at: dormant.appendingPathComponent("aura"),
                                            withIntermediateDirectories: true)
            for entry in mirroredContractEntries {
                let src = active.appendingPathComponent(entry)
                let dst = dormant.appendingPathComponent(entry)
                if fileManager.fileExists(atPath: dst.path) {
                    try fileManager.removeItem(at: dst)
                }
                if fileManager.fileExists(atPath: src.path) {
                    try fileManager.copyItem(at: src, to: dst)
                }
            }
        }
    }
}

File: studio/windows/AuraStudio.Core/Installer/FirmwareSwitcher.cs
using System.Globalization;
using System.Text;
using AuraStudio.Core;

namespace AuraStudio.Core.Installer;

/// <summary>
/// ST-056 / <c>CONTRATO-firmware-studio.md</c> v10 (v14: hasta tres familias):
/// varios firmwares instalados a la vez y conmutación entre ellos POR RENOMBRE,
/// sin borrar ni descargar. Port de <c>Services/FirmwareSwitcher.swift</c>.
///
/// El árbol activo es siempre <c>/.rockbox/</c> (lo único que el bootloader
/// compartido sabe arrancar); el de cualquier otra familia duerme como
/// <c>/.firmware-&lt;familia&gt;/</c>, completo y con sus propios ajustes. Cambiar
/// de firmware son dos renombres en FAT (instantáneos), en un orden que importa
/// — el saliente primero — más el respaldo del bootloader en la raíz y el
/// marcador de sync para que el entrante reconstruya su base de datos.
///
/// <b>INVARIANTE (ST-069/ST-073/ST-147, contrato v15/v16/v19)</b>: todo lo de
/// acá opera SOLO sobre <c>/.rockbox/</c>, <c>/.firmware-*/</c> y
/// <c>/rockbox.ipod</c>. <c>/.aura/tagcache/</c> (base compartida por los tres
/// firmwares), <c>/.aura/thumbs/</c> (miniaturas compartidas),
/// <c>/.aura/art/</c> (caché maestra de imágenes) y
/// <see cref="SharedSettingsRelativePath"/> (ajustes compartidos) son
/// propiedad del firmware aunque vivan bajo <c>/.aura/</c>: ningún renombre,
/// estacionamiento, reparación, siembra ni espejo los toca.
/// </summary>
public static class FirmwareSwitcher
{
    public const string ActiveTreeName = ".rockbox";
    public const string RootFirmwareBinaryName = "rockbox.ipod";

    /// <summary>
    /// ST-147 / contrato v19: ajustes compartidos entre las tres familias
    /// (bloqueo, brillo, idioma, etc.). Es un ARCHIVO, no un directorio, pero
    /// la regla es la misma que <c>/.aura/art/</c> desde ST-073 — Studio nunca
    /// lo borra, mueve ni reescribe. Ningún código de este repo enumera
    /// <c>/.aura/</c> de forma amplia (cada operación nombra explícitamente lo
    /// que le toca, ver <c>LibrarySyncEngine.ClearFirmwareDatabases</c>), así
    /// que esta constante existe para que la protección quede fijada por
    /// pruebas, no solo asumida.
    /// </summary>
    public const string SharedSettingsRelativePath = ".aura/settings.cfg";

    /// <summary>
    /// ST-059 / contrato v12: sello de biblioteca compartido — cambia solo
    /// cuando un sync toca música. Igualdad exacta de sellos = la base del árbol
    /// entrante sigue válida = no se escribe marcador.
    /// </summary>
    public const string LibraryStampRelativePath = ".aura/library-stamp";

    /// <summary>
    /// Hasta v14 la anotación vivía por árbol; desde v15 el firmware la
    /// lee/escribe en <c>/.aura/tagcache/db_stamp.txt</c>. Studio conserva la
    /// ruta por árbol solo para compararla al cambiar de familia con árboles
    /// anteriores a v15.
    /// </summary>
    public const string DbStampRelativePathInTree = "aura/db_stamp.txt";

    /// <summary>Motivos por los que un cambio de familia no se puede hacer.</summary>
    public enum SwitchFailure
    {
        None,
        /// <summary>No hay <c>/.firmware-&lt;familia&gt;/</c> que despertar.</summary>
        DormantTreeMissing,
        /// <summary>La familia pedida ya es la activa.</summary>
        AlreadyActive,
        /// <summary>Una familia que no se puede estacionar ni despertar (desconocida).</summary>
        FamilyNotSwitchable
    }

    public sealed class SwitchException(SwitchFailure failure, FirmwareFamily family)
        : Exception(Describe(failure, family))
    {
        public SwitchFailure Failure { get; } = failure;
        public FirmwareFamily Family { get; } = family;

        private static string Describe(SwitchFailure failure, FirmwareFamily family) => failure switch
        {
            SwitchFailure.DormantTreeMissing => $"No hay un árbol dormido de {family.DisplayName} en este iPod.",
            SwitchFailure.AlreadyActive => $"{family.DisplayName} ya es el firmware activo.",
            SwitchFailure.FamilyNotSwitchable => $"Aura Studio no sabe cómo estacionar ni despertar {family.DisplayName}.",
            _ => "No se pudo cambiar de firmware."
        };
    }

    private static string Combine(string root, string relative) =>
        Path.Combine(root, relative.Replace('/', Path.DirectorySeparatorChar));

    /// <summary>Qué familias tienen un árbol dormido en el volumen.</summary>
    public static IReadOnlyList<FirmwareFamily> DormantFamilies(string volumeRoot) =>
        FirmwareFamily.Installable
            .Where(family => family.DormantTreeName is { } name && Directory.Exists(Path.Combine(volumeRoot, name)))
            .ToList();

    public static bool HasActiveTree(string volumeRoot) =>
        Directory.Exists(Path.Combine(volumeRoot, ActiveTreeName));

    /// <summary>
    /// Un cambio que quedó a medias (batería, cable): sin <c>/.rockbox/</c> pero
    /// con exactamente un árbol dormido. Se repara despertándolo. Con dos
    /// dormidos y ninguno activo <b>no se adivina</b>: se deja como está y se
    /// devuelve <c>null</c> (contrato v14 — Studio pide elegir).
    /// </summary>
    public static FirmwareFamily? RepairIfNeeded(string volumeRoot)
    {
        if (HasActiveTree(volumeRoot)) return null;
        var dormant = DormantFamilies(volumeRoot);
        if (dormant.Count != 1) return null;
        var family = dormant[0];
        if (family.DormantTreeName is not { } name) return null;

        Directory.Move(Path.Combine(volumeRoot, name), Path.Combine(volumeRoot, ActiveTreeName));
        RefreshRootBinary(volumeRoot);
        return family;
    }

    /// <summary>
    /// Estaciona el árbol activo como dormido de <paramref name="family"/>
    /// (reemplazando un dormido anterior de esa misma familia, si lo hubiera —
    /// nunca dos de la misma). Lo usa el instalador antes de instalar OTRA
    /// familia en <c>/.rockbox/</c>, en vez de borrarla.
    /// </summary>
    public static void ParkActiveTree(FirmwareFamily family, string volumeRoot)
    {
        if (family.DormantTreeName is not { } name)
            throw new SwitchException(SwitchFailure.FamilyNotSwitchable, family);

        string active = Path.Combine(volumeRoot, ActiveTreeName);
        string dormant = Path.Combine(volumeRoot, name);
        if (!Directory.Exists(active)) return;
        if (Directory.Exists(dormant)) Directory.Delete(dormant, true);
        Directory.Move(active, dormant);
    }

    /// <summary>
    /// Borra el árbol dormido de <paramref name="family"/> si existe (tras
    /// instalar esa familia fresca en <c>/.rockbox/</c>: nunca un dormido de la
    /// familia activa).
    /// </summary>
    public static void RemoveDormantTree(FirmwareFamily family, string volumeRoot)
    {
        if (family.DormantTreeName is not { } name) return;
        string dormant = Path.Combine(volumeRoot, name);
        if (Directory.Exists(dormant)) Directory.Delete(dormant, true);
    }

    /// <summary>
    /// EL cambio (contrato v10, pasos 2–5; el 1 — que el firmware guarde lo suyo
    /// — no aplica desde Studio porque el firmware no está corriendo en modo
    /// disco, y el 6 — reiniciar — lo hace el usuario al expulsar):
    /// <code>
    /// /.rockbox/               -> /.firmware-&lt;activa&gt;/
    /// /.firmware-&lt;destino&gt;/ -> /.rockbox/
    /// /rockbox.ipod            := /.rockbox/rockbox.ipod del entrante
    /// /.aura/sync-pending.json  con music: true, SOLO si el sello lo pide (v12)
    /// </code>
    /// </summary>
    public static void SwitchActiveFirmware(FirmwareFamily target, FirmwareFamily currentlyActive, string volumeRoot)
    {
        if (Equals(target, currentlyActive)) throw new SwitchException(SwitchFailure.AlreadyActive, target);
        if (target.DormantTreeName is not { } targetName) throw new SwitchException(SwitchFailure.FamilyNotSwitchable, target);
        if (currentlyActive.DormantTreeName is not { } activeName) throw new SwitchException(SwitchFailure.FamilyNotSwitchable, currentlyActive);

        string active = Path.Combine(volumeRoot, ActiveTreeName);
        string targetDormant = Path.Combine(volumeRoot, targetName);
        string parkedActive = Path.Combine(volumeRoot, activeName);

        if (!Directory.Exists(targetDormant)) throw new SwitchException(SwitchFailure.DormantTreeMissing, target);

        // (2) saliente primero: el peor caso (corte acá) deja un árbol dormido
        // entero y ninguno activo — RepairIfNeeded() lo levanta.
        if (Directory.Exists(active))
        {
            if (Directory.Exists(parkedActive)) Directory.Delete(parkedActive, true);
            Directory.Move(active, parkedActive);
        }
        // (3) entrante
        Directory.Move(targetDormant, active);
        // (4) respaldo del bootloader
        RefreshRootBinary(volumeRoot);
        // (5) v12/ST-059: el marcador solo si la biblioteca cambió desde que el
        // ENTRANTE construyó su base — sin sync de por medio, el cambio es
        // instantáneo y sin reconstrucción.
        if (!IncomingDatabaseIsCurrent(volumeRoot, currentlyActive))
        {
            new SyncPendingMarker(new SyncPendingMarker.Changes(Music: true, Video: false, Images: false))
                .Write(volumeRoot);
        }

        // ST-146 / maestro §B: el árbol que acaba de despertar puede llevar
        // días o semanas dormido — su reloj queda tan atrasado como el último
        // apagado. Se sincroniza acá, en el árbol ENTRANTE (ya renombrado a
        // ActiveTreeName arriba), sin esperar a que el usuario desconecte y
        // vuelva a conectar.
        ClockSyncWriter.WriteToDisk(volumeRoot);
    }

    /// <summary>
    /// ST-059: compara el sello del árbol ya renombrado a <c>/.rockbox/</c> (el
    /// entrante) contra <c>/.aura/library-stamp</c>. Si el sello compartido falta
    /// (primer cambio tras v12), lo crea y lo anota como del SALIENTE — su base
    /// sí está al día, acaba de estar corriendo — para que el próximo cambio de
    /// vuelta ya no reconstruya.
    /// </summary>
    private static bool IncomingDatabaseIsCurrent(string volumeRoot, FirmwareFamily outgoing)
    {
        string stampPath = Combine(volumeRoot, LibraryStampRelativePath);
        string? stamp = ReadTrimmed(stampPath);

        if (string.IsNullOrEmpty(stamp))
        {
            string fresh = MakeLibraryStamp();
            TryWriteLine(stampPath, fresh);
            if (outgoing.DormantTreeName is { } name)
            {
                TryWriteLine(Path.Combine(volumeRoot, name, DbStampRelativePathInTree.Replace('/', Path.DirectorySeparatorChar)), fresh);
            }
            stamp = fresh;
        }

        string incoming = Path.Combine(volumeRoot, ActiveTreeName,
            DbStampRelativePathInTree.Replace('/', Path.DirectorySeparatorChar));
        string? recorded = ReadTrimmed(incoming);
        return recorded is not null && recorded == stamp;
    }

    /// <summary>Una línea opaca: marca de tiempo ISO-8601 + sufijo aleatorio. Se compara por igualdad exacta, nunca se interpreta.</summary>
    public static string MakeLibraryStamp(DateTimeOffset? date = null) =>
        (date ?? DateTimeOffset.Now).ToString("yyyy-MM-ddTHH:mm:sszzz", CultureInfo.InvariantCulture)
        + "-" + Guid.NewGuid().ToString("N")[..8];

    /// <summary>
    /// ST-059: Studio renueva el sello en cada sync que toca música — es LA
    /// definición de "la biblioteca cambió" para el cambio de firmware.
    /// </summary>
    public static void BumpLibraryStamp(string volumeRoot) =>
        TryWriteLine(Combine(volumeRoot, LibraryStampRelativePath), MakeLibraryStamp());

    /// <summary>
    /// <c>/rockbox.ipod</c> (raíz) = el binario del árbol activo. Es lo que el
    /// bootloader arranca si <c>/.rockbox/rockbox.ipod</c> falta; tiene que ser
    /// SIEMPRE el del firmware activo, nunca el de un árbol dormido.
    /// </summary>
    public static void RefreshRootBinary(string volumeRoot)
    {
        string source = Path.Combine(volumeRoot, ActiveTreeName, RootFirmwareBinaryName);
        string root = Path.Combine(volumeRoot, RootFirmwareBinaryName);
        if (!File.Exists(source)) return;
        File.Copy(source, root, overwrite: true);
    }

    /// <summary>
    /// Archivos del contrato que Studio escribe en <c>.rockbox/aura/</c> y que el
    /// firmware dormido también necesita al despertar (v10). <c>aura.cfg</c> NO
    /// (es de cada firmware); <c>themes/</c> tampoco (viaja con su árbol).
    /// <c>aura/db_stamp.txt</c> tampoco: es la anotación de CADA árbol sobre su
    /// propia base. <c>aura/install_manifest.cfg</c> tampoco (v11): es POR ÁRBOL
    /// y espejarlo haría que la actualización selectiva del dormido diera por
    /// escritos archivos que nunca se le escribieron.
    /// </summary>
    public static readonly IReadOnlyList<string> MirroredContractEntries =
    [
        "aura/sync_summary.cfg",
        "aura/sync_manifest.json",
        "aura/artist_images.cfg",
        "aura/artists",
        "aura/video_categories.cfg",
        "aura/photo_categories.cfg",
        "aura/ratings.cfg",
        "aura/device.cfg"
    ];

    /// <summary>
    /// ST-061: el problema inverso al espejo — un árbol ACTIVO recién instalado
    /// (extracción fresca) no tiene los archivos del contrato que Studio escribió
    /// en su vida anterior: sin <c>sync_summary.cfg</c> el firmware dice "sin
    /// sincronizar todavía", sin <c>artist_images.cfg</c>/<c>artists/</c> no hay
    /// fotos de artista, sin <c>*_categories.cfg</c> los videos/fotos salen sin
    /// clasificar. Si el árbol dormido SÍ los tiene, se copian de ahí. Solo
    /// cuando el activo NO tiene <c>sync_summary.cfg</c> — si lo tiene, lo suyo
    /// es igual o más nuevo.
    /// </summary>
    public static bool SeedContractFilesToActiveTree(string volumeRoot)
    {
        string active = Path.Combine(volumeRoot, ActiveTreeName);
        if (File.Exists(Combine(active, "aura/sync_summary.cfg"))) return false;

        var donor = DormantFamilies(volumeRoot)
            .Select(family => family.DormantTreeName)
            .OfType<string>()
            .Select(name => Path.Combine(volumeRoot, name))
            .FirstOrDefault(dir => File.Exists(Combine(dir, "aura/sync_summary.cfg")));
        if (donor is null) return false;

        Directory.CreateDirectory(Path.Combine(active, "aura"));
        foreach (var entry in MirroredContractEntries)
        {
            CopyEntry(Combine(donor, entry), Combine(active, entry));
        }
        return true;
    }

    /// <summary>Espeja los archivos del contrato del árbol activo a TODOS los dormidos presentes (v10/v14).</summary>
    public static void MirrorContractFilesToDormantTrees(string volumeRoot)
    {
        string active = Path.Combine(volumeRoot, ActiveTreeName);
        foreach (var family in DormantFamilies(volumeRoot))
        {
            if (family.DormantTreeName is not { } name) continue;
            string dormant = Path.Combine(volumeRoot, name);
            Directory.CreateDirectory(Path.Combine(dormant, "aura"));
            foreach (var entry in MirroredContractEntries)
            {
                CopyEntry(Combine(active, entry), Combine(dormant, entry));
            }
        }
    }

    /// <summary>Copia un archivo o un directorio completo, reemplazando el destino.</summary>
    private static void CopyEntry(string source, string destination)
    {
        if (Directory.Exists(source))
        {
            if (Directory.Exists(destination)) Directory.Delete(destination, true);
            Directory.CreateDirectory(destination);
            foreach (string file in Directory.EnumerateFiles(source, "*", SearchOption.AllDirectories))
            {
                string target = Path.Combine(destination, Path.GetRelativePath(source, file));
                Directory.CreateDirectory(Path.GetDirectoryName(target)!);
                File.Copy(file, target, overwrite: true);
            }
            return;
        }

        if (!File.Exists(source)) return;
        if (Directory.Exists(destination)) Directory.Delete(destination, true);
        Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
        File.Copy(source, destination, overwrite: true);
    }

    private static string? ReadTrimmed(string path)
    {
        try
        {
            return File.Exists(path) ? File.ReadAllText(path, Encoding.UTF8).Trim() : null;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }

    private static void TryWriteLine(string path, string value)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            File.WriteAllText(path, value + "\n", new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            // Perder el sello degrada a "reconstruir la base una vez de más";
            // fallar el cambio de firmware entero sería mucho peor.
        }
    }
}
